### PR TITLE
8261744: Implement CharsetDecoder ASCII and latin-1 fast-paths

### PR DESCRIPTION
**Pull request by cl4es**
This patch exposes a couple of intrinsics used by String to speed up ASCII checking and byte[] -> char[] inflation, which can be used by latin1 and ASCII-compatible CharsetDecoders to speed up decoding operations.

- Fast-path implemented for all standard charsets, with up to 10x performance improvements in microbenchmarks reading Strings from ByteArrayInputStream. 
- Cleanup of StreamDecoder/-Encoder with some minor improvements when interpreting
- Reworked creation of JavaLangAccess to be safely published for CharsetDecoders/-Encoders used for setting up System.out/in. As JLA and these encoders are created during System.initPhase1 the current sequence caused the initialization to became unstable and a few tests were consistently getting an NPE.

Testing: tier1-3

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261744](https://bugs.openjdk.java.net/browse/JDK-8261744): Implement CharsetDecoder ASCII and latin-1 fast-paths

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2574/head:pull/2574`
`$ git checkout pull/2574`

### CODE REVIEW

**Comment by cl4es** [comment]
On the best-case microbenchmark (the byte stream is all ASCII), speed-ups are quite telling:

Before:
```
ByteStreamDecoder.readStringReader                          US-ASCII       256  avgt   10    2085.399 ±    66.522  ns/op
ByteStreamDecoder.readStringReader                          US-ASCII      4096  avgt   10   12466.702 ±   747.116  ns/op
ByteStreamDecoder.readStringReader                          US-ASCII     25000  avgt   10  123508.987 ±  3583.345  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-1       256  avgt   10    1894.185 ±    51.772  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-1      4096  avgt   10    8117.404 ±   594.708  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-1     25000  avgt   10   99409.792 ± 28308.936  ns/op
ByteStreamDecoder.readStringReader                             UTF-8       256  avgt   10    2090.337 ±    56.500  ns/op
ByteStreamDecoder.readStringReader                             UTF-8      4096  avgt   10   11698.221 ±   898.910  ns/op
ByteStreamDecoder.readStringReader                             UTF-8     25000  avgt   10   66568.987 ±  4204.361  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-6       256  avgt   10    3061.130 ±   120.132  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-6      4096  avgt   10   24623.494 ±  1916.362  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-6     25000  avgt   10  139138.140 ±  7109.636  ns/op
ByteStreamDecoder.readStringReader                             MS932       256  avgt   10    2612.535 ±    98.638  ns/op
ByteStreamDecoder.readStringReader                             MS932      4096  avgt   10   18843.438 ±  1767.822  ns/op
ByteStreamDecoder.readStringReader                             MS932     25000  avgt   10  119923.997 ± 18560.065  ns/op
```

After:
```
ByteStreamDecoder.readStringReader                          US-ASCII       256  avgt   10    1556.588 ±    37.083  ns/op
ByteStreamDecoder.readStringReader                          US-ASCII      4096  avgt   10    3290.627 ±   125.327  ns/op
ByteStreamDecoder.readStringReader                          US-ASCII     25000  avgt   10   13118.794 ±   597.086  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-1       256  avgt   10    1525.460 ±    36.510  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-1      4096  avgt   10    3051.887 ±   113.036  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-1     25000  avgt   10   11401.228 ±   563.124  ns/op
ByteStreamDecoder.readStringReader                             UTF-8       256  avgt   10    1596.878 ±    43.824  ns/op
ByteStreamDecoder.readStringReader                             UTF-8      4096  avgt   10    3349.961 ±   119.278  ns/op
ByteStreamDecoder.readStringReader                             UTF-8     25000  avgt   10   13273.403 ±   591.600  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-6       256  avgt   10    1602.328 ±    44.092  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-6      4096  avgt   10    3403.312 ±   107.516  ns/op
ByteStreamDecoder.readStringReader                        ISO-8859-6     25000  avgt   10   13163.468 ±   709.642  ns/op
ByteStreamDecoder.readStringReader                             MS932       256  avgt   10    1602.837 ±    32.021  ns/op
ByteStreamDecoder.readStringReader                             MS932      4096  avgt   10    3379.439 ±    87.716  ns/op
ByteStreamDecoder.readStringReader                             MS932     25000  avgt   10   13376.980 ±   669.983  ns/op
```

Performance degrades when you mix in non-ASCII characters, but so does the `new String(byte[], Charset)` baseline. While there might be algorithmic refinements possible to improve on some of the non-ASCII variants I'm happy to leave that to a follow-up and keep this RFE simple.

**Comment by marschall** [comment]
Is there a reason `sun.nio.cs.ISO_8859_1.Encoder#implEncodeISOArray(char[], int, byte[], int, int)` wasn't moved to `JavaLangAccess` as well?

**Comment by cl4es** [comment]
> Is there a reason `sun.nio.cs.ISO_8859_1.Encoder#implEncodeISOArray(char[], int, byte[], int, int)` wasn't moved to `JavaLangAccess` as well?

Exposing StringUTF16.compress for Latin-1 and ASCII-compatible encoders seem very reasonable, which I was thinking of exploring next as a separate RFE.

**Comment by cl4es** [comment]
> > Is there a reason `sun.nio.cs.ISO_8859_1.Encoder#implEncodeISOArray(char[], int, byte[], int, int)` wasn't moved to `JavaLangAccess` as well?
> 
> Exposing StringUTF16.compress for Latin-1 and ASCII-compatible encoders seem very reasonable, which I was thinking of exploring next as a separate RFE.

Maybe I misunderstood. The intrinsified method you point out here pre-dates the work in JDK 9 to similarly intrinsify char[]->byte[] compaction in StringUTF16, see https://bugs.openjdk.java.net/browse/JDK-6896617

It might be worthwhile cleaning this up. Not having to route via SharedSecrets -> JavaLangAccess does speed things up during startup/interpretation, at the cost of some code duplication.

**Comment by AlanBateman** [comment]
This looks a really good change and it's great can these decoders get the benefit of the instrinics work.

ISO_8869_1.decodeArrayLoop using JLA.inflate looks a bit strange and maybe we can rename the method in the shared secret to inflatedCopy or just copy. or just add a comment. The reason is that code outside of java.lang shouldn't know anything about String's internals and inflation.

The reformatting changes to StreamEncoder/StreamDecoder make it hard to see if there are any actual changes. Is there anything we should look at? It's okay to include this cleanup, I can't tell what happened with this source file.

I want to study the change System.setJavaLangAccess a bit further to understand why this is needed (I know there is a awkward bootstrapping issue here).

I expect Naoto will want to review this PR too.

**Comment by cl4es** [comment]
I added the `Inflated copy from byte[] to char[], as defined by StringLatin1` comment to the `inflate` method in `JavaLangAccess`, but you're probably right and it would be good to make the name more explicit when exporting it outside of the package internal use. How about `inflateBytesToChars`?

**Comment by cl4es** [comment]
> The reformatting changes to StreamEncoder/StreamDecoder make it hard to see if there are any actual changes. Is there anything we should look at? It's okay to include this cleanup, I can't tell what happened with this source file.
> 

Apart from a few things being made final that weren't, and a flattening of a back-to-back Charset lookup in one of the constructors its mostly formatting cleanups.

> I want to study the change System.setJavaLangAccess a bit further to understand why this is needed (I know there is a awkward bootstrapping issue here).

Right, I'm not exactly sure why the more limited changes I attempted in 5f4e87f failed. In that change I simply changed the initialization order, which made the failing (closed) tests pass locally - but not in our CI. Since this is in initPhase1 there should be no concurrency possible.

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -1008,6 +1008,26 @@ private static boolean isASCII(byte[] src) {
 
     //////////////////////////////// utf8 ////////////////////////////////////
 
+    /**
+     * Decodes ASCII from the source byte array into the destination
+     * char array.
+     *
+     * @return the number of bytes successfully decoded, at most len
+     */
+    /* package-private */
```
Comment (naotoj) [inline src/java.base/share/classes/java/lang/String.java:1017]: Some more explanation would be helpful here, as it is accessed from System for shared secrets.

File: src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java [hunk 2]
```diff
@@ -45,6 +45,7 @@
 
 import jdk.internal.module.ServicesCatalog;
 import jdk.internal.reflect.ConstantPool;
+import jdk.internal.vm.annotation.IntrinsicCandidate;
```
Comment (naotoj) [inline src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java:48]: Not needed.

**Review by naotoj — COMMENTED** [review]
Great improvement! Looks good to me overall. I wondered about the changes in JLA as Alan already mentioned.

File: src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java (hunk 2 above)
Comment (cl4es) [reply src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java]: Fixed

File: src/java.base/share/classes/java/lang/String.java (hunk 1 above)
Comment (cl4es) [reply src/java.base/share/classes/java/lang/String.java:1018]: Ok, added a short comment

**Comment by AlanBateman** [comment]
> Right, I'm not exactly sure why the more limited changes I attempted in [5f4e87f](https://github.com/openjdk/jdk/commit/5f4e87f50f49e64b8616063c176ea35632b0347e) failed. In that change I simply changed the initialization order, which made the failing (closed) tests pass locally - but not in our CI. Since this is in initPhase1 there should be no concurrency possible.

The Reference Handler thread is started by the initializer in jl.ref.Reference so could be a candidate. The Finalizer thread is another but this should VM.awaitInitLevel(1) and not touch JLA until initPhase1 is done.

**Comment by AlanBateman** [comment]
> but you're probably right and it would be good to make the name more explicit when exporting it outside of the package internal use. How about `inflateBytesToChars`?

That should be okay.

**Comment by marschall** [comment]
> > > Is there a reason `sun.nio.cs.ISO_8859_1.Encoder#implEncodeISOArray(char[], int, byte[], int, int)` wasn't moved to `JavaLangAccess` as well?
> > 
> > 
> > Exposing StringUTF16.compress for Latin-1 and ASCII-compatible encoders seem very reasonable, which I was thinking of exploring next as a separate RFE.
> 
> Maybe I misunderstood. The intrinsified method you point out here pre-dates the work in JDK 9 to similarly intrinsify char[]->byte[] compaction in StringUTF16, see https://bugs.openjdk.java.net/browse/JDK-6896617
> 
> It might be worthwhile cleaning this up. Not having to route via SharedSecrets -> JavaLangAccess does speed things up during startup/interpretation, at the cost of some code duplication.

My understanding was ISO_8859_1$Encoder.implEncodeISOArray and StringUTF16.compress are ultimately hooked up to the same intrinsic. I find it inconsistent that ISO_8859_1$Encoder access an encoding intrinsitc directly while ISO_8859_1$Decoder and others access a decoding intrinsic indirectly through JavaLangAccess. I realize this RFE is about decoding so keeping encoding to a different RFE may indeed be better.

**Comment by cl4es** [comment]
> > Right, I'm not exactly sure why the more limited changes I attempted in [5f4e87f](https://github.com/openjdk/jdk/commit/5f4e87f50f49e64b8616063c176ea35632b0347e) failed. In that change I simply changed the initialization order, which made the failing (closed) tests pass locally - but not in our CI. Since this is in initPhase1 there should be no concurrency possible.
> 
> The Reference Handler thread is started by the initializer in jl.ref.Reference so could be a candidate. The Finalizer thread is another but this should VM.awaitInitLevel(1) and not touch JLA until initPhase1 is done.

Turns out the native code called by `SystemProps.initProperties();` can initialize a `CharsetDecoder` if `sun.jnu.encoding` is set to certain values, and this tripped up the initialization to cause `getJavaLangAccess` to be called before `setJavaLangAccess`. By moving `setJavaLangAccess` to the very start of `initPhase1` we avoid this possibility, and limit churn.

**Comment by AlanBateman** [comment]
> Turns out the native code called by `SystemProps.initProperties();` can initialize a `CharsetDecoder` if `sun.jnu.encoding` is set to certain values, and this tripped up the initialization to cause `getJavaLangAccess` to be called before `setJavaLangAccess`. By moving `setJavaLangAccess` to the very start of `initPhase1` we avoid this possibility, and limit churn.

I assume this is tests running with -Dfile.encoding. I can go along with moving setJavaLangAccess to the start of initPhase1. I'll do another pass over the updated patch today. Good work!

**Review by AlanBateman — APPROVED** [review]

**Review by naotoj — APPROVED** [review]

**Comment by cl4es** [comment]
Thanks for reviewing!

/integrate

**Pull request closed without merging**